aggregator_core/src/datastore/test_util.rs

```diff
@@ -49,24 +59,73 @@ impl EphemeralDatabase {
+        let (port_tx, port_rx) = oneshot::channel();
+
+        // Hack: run testcontainer logic under its own thread with its own tokio runtime, to avoid
+        // deadlocking the main runtime when waiting for logs to finish in the Drop implementation.
+        let db_thread = std::thread::spawn(move || {
+            tokio::runtime::Builder::new_current_thread()
+                .enable_all()
+                .build()
+                .unwrap()
+                .block_on(async move {
+                    // Start an instance of Postgres running in a container.
+                    let db_container = ContainerRequest::from(Postgres::default())
+                        .with_cmd(Self::postgres_configuration(&[
+                            // Many tests running concurrently can overwhelm the available,
+                            // connections, so bump the limit.
+                            "max_connections=200",
+                            // Enable logging of query plans.
+                            "shared_preload_libraries=auto_explain",
+                            "log_min_messages=LOG",
+                            "auto_explain.log_min_duration=0",
+                            "auto_explain.log_analyze=true",
+                            // Discourage postgres from doing sequential scans, so we can analyze
+                            // whether we have appropriate indexes in unit tests. Because test
+                            // databases are not seeded with data, the query planner will often
+                            // choose a sequential scan because it would be faster than hitting an
+                            // index.
+                            "enable_seqscan=false",
+                        ]))
+                        .start()
+                        .await
+                        .unwrap();
+
+                    let stdout = db_container.stdout(true);
+                    let stderr = db_container.stderr(true);
+                    let log_consumer_handle =
+                        env::var_os("JANUS_TEST_DUMP_POSTGRESQL_LOGS").map(|_| {
+                            tokio::spawn(async move {
+                                join!(buffer_printer(stdout), buffer_printer(stderr));
+                            })
+                        });
+
+                    const POSTGRES_DEFAULT_PORT: u16 = 5432;
+                    let port_number = db_container
+                        .get_host_port_ipv4(POSTGRES_DEFAULT_PORT)
+                        .await
+                        .unwrap();
+                    trace!("Postgres container is up with port {port_number}");
+
+                    // Send port information, which frees this function to continue.
+                    port_tx.send(port_number).unwrap();
+
+                    // Wait for shutdown to be signalled.
+                    shutdown_rx.await.unwrap();
+                    drop(db_container);
```

Comment:
I'd recommend trying out calling `ContainerAsync::stop()` first, then waiting for the log consumers to finish, then dropping to delete the container. If the log consumer streams stop once the container has stopped, then that should work , and it will prevent race conditions leading to lost logs.

Reply:
Nice, that appears to work and is much cleaner, insofar as I still see the `database system is shut down` at the end of logs.